keps/sig-api-machinery/NNNN-watch-list/README.md

```diff
@@ -0,0 +1,956 @@
+We do that on every ListAndWatch call (init or relist). 
+Each event (ADD, UPDATE, DELETE) except the BOOKMARK event received from the server is collected.
+Passing resourceVersionMatch=ConsistentInitialAdds tells the cacher it has to guarantee that the cache is at least up to date as a LIST executed at the same time.
+
+Note: Unfortunately as of today, the watch cache is vulnerable to stale reads, see https://github.com/kubernetes/kubernetes/issues/59848 for more details.
+
+Note 2: This preserves the current behavior of relisting on every ListAndWatch call.
+
+Step 2: Right after receiving a request from the reflector, the cacher gets the current resourceVersion (aka bookmarkAfterResourceVersion) directly from the etcd.
+It is used to make sure the cacher is up to date (has seen data stored in etcd) and to let the reflector know it has seen all initial data.
+There are ways to do that cheaply, e.g. we could issue a count request against the datastore.
+Next, the cacher creates a new cacheWatcher (implements watch.Interface) passing the given bookmarkAfterResourceVersion, and gets initial data from the watchCache.
+After sending initial data the cacheWatcher starts listening on an incoming channel for new events, including a bookmark event.
+At some point, the cacher will receive an event with the resourceVersion equal or greater to the bookmarkAfterResourceVersion.
+It will be propagated to the cacheWatcher and then back to the reflector as a BOOKMARK event.
+
+![watch cache with the proposed changes](./cacher.png "watch cache with the proposed changes")
+
+Step 2a: Where does the initial data come from?
+
+During construction, the cacher creates the reflector and the watchCache.
+Since the watchCache implements the Store interface it is used by the reflector to store all data it has received from etcd.
+Note: Right now the reflector used to populate the watchCache uses the old semantics of LIST/WATCH.
+This could be optimized in the future (but out-of-scope here) by establishing a WATCH directly with etcd and waiting for the bookmarkAfterResourceVersion.
+
+Step 2b: What happens when new events are received while the cacheWatcher is sending initial data?
+
+The cacher maintains a list of all current watchers (cacheWatcher) and a separate goroutine (dispatchEvents) for delivering new events to the watchers.
+New events are added via the cacheWatcher.nonblockingAdd method that adds an event to the cacheWatcher.incoming channel.
+The cacheWatcher.incoming is a buffered channel and has a different size for different Resources (10 or 1000).
+Since the cacheWatcher starts processing the cacheWatcher.incoming channel only after sending all initial events it might block once its buffered channel tips over.
+In that case, it will be added to the list of blockedWatchers and will be given another chance to deliver an event after all nonblocking watchers have sent the event.
+All watchers that have failed to deliver the event will be closed.
```

Comment:
Where we will land effectively is that if we won't be able to send the "initial state" before <size of buffer>+1 events will be delivered to the watch in the meantime, the watcher will get closed.

We may try to play a bit with the buffer size, to mitigate the impact of that.

Reply:
I think we should be fine. The size of the buffer is different for different resources but in the range of 10 or 1000. In the worst-case scenario, we restart the WATCH request.